Better ordering data structure

RaduBerinde opened the pull request:
The current ordering code is hard to understand (the logic of what the []int
actually represents is spread around the codebase) and there are some problems
with the representation, outlined in issue #3914.

In this change we introduce a better representation and abstraction for ordering
information and refactor the code accordingly.

@andreimatei @petermattis 
This should fix the example query in #3914 - I will add some tests like that to verify.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3918)

<!-- Reviewable:end -->

RaduBerinde committed:
Better ordering data structure

The current ordering code is hard to understand (the logic of what the []int
actually represents is spread around the codebase) and there are some problems
with the representation, outlined in issue #3914.

In this change we introduce a better representation and abstraction for ordering
information and refactor the code accordingly.

petermattis commented:
LGTM, though we need those additional tests showing how this avoids sorting.

---

Review status: 0 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/ordering.go, line 19 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8PrRJ8Nto9FqRCVEJT:-K8PrRJ8Nto9FqRCVEJU:61047814) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L19)):*
`s/desiredOrdering/columnOrdering/g`

---

*[sql/ordering.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8PqgnYlrZ1nMJCWH35:-K8PqgnZLFYZS0cG0f6X:90853083) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L48)):*
I'm finding myself confused by this name. How about `exactMatchCols`?

---

*[sql/ordering_test.go, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8PsF8Z1SL_1gsgshnr:-K8PsF8Z1SL_1gsgshns:-1611140564) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering_test.go#L26)):*
FYI, you can actually define the `desiredCase` and `computeOrderCase` types within the `TestComputeOrderMatch` function. If you don't need to use these types outside of that function this is preferable. The only downside is that you can't define functions on types defined within a method.

---

*[sql/ordering_test.go, line 61 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8Psfl_zO_uutlKG-79:-K8Psfl_zO_uutlKG-7A:1416022955) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering_test.go#L61)):*
You might define a smaller helper function to help with these test case definitions:

```
  defTestCase := func(expected, expectedReverse int, desired columnOrdering) desiredCase {
    _, _, line, _ := runtime.Caller(2)
    return desiredCase{
      line: line,
      desired: desired,
      expected: expected,
      expectedReverse: expectedReverse,
    }
  }
```

And then to use:

```
  cases: []desiredCase{
    defTestCase(0, 0, columnOrdering({1, true}, {5, false})),
  },
```

---

*[sql/values.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8Pu5IJLFELC77jElJN:-K8Pu5IKVmnr5tPgNA7N:-87443626) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/values.go#L89)):*
I think you can return `nil` here. An empty slice and a nil slice are usually equivalent.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-:-K8PuFn56yS6DMXqJTsW:-1499966486)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[sql/distinct.go, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e-K310S1LQK9HGr_X:-K8e-K32Zz2Sh7WHz3Q9:833488553) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/distinct.go#L37)):*
`isNil` is odd. can you make `Ordering` return `(orderingInfo, bool)` instead? Ah, I see @petermattis also suggested that you simply return a `nil` `orderingInfo`, which I think is a better idea.

---

*[sql/distinct.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e2tND0x6FGbuyrACA:-K8e2tNEwPVyI-6Tj9c8:1029120080) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/distinct.go#L40)):*
"ignore it" might be technically true, but it is not consistent with the terminology used here. How about "we can consider it as being in order"?

---

*[sql/ordering.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8PqgnYlrZ1nMJCWH35:-K8dzXyIVTe8B3CC40YR:-647422117) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L48)):*
i think the word "single" is important here, because an index on `(a,b,c,d)` with a `WHERE a IN (1,2,3)` would also imply that `a` is an `exactMatchCol`

---

*[sql/ordering.go, line 71 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e-rLl2757t-8EQrol:-K8e-rLl2757t-8EQrom:-715881088) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L71)):*
nit: how come this isn't a method?

---

*[sql/ordering.go, line 79 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e0PwrJ-VxENXOOaOM:-K8e0PwrJ-VxENXOOaON:-102234440) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L79)):*
nit: do you think this would be clearer as `(ci.reverse == col.reverse) != reverse`?

---

*[sql/ordering_test.go, line 38 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e1FDw047J0oAvILGm:-K8e1FDxx6wCBnMgurTx:852958883) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering_test.go#L38)):*
s/indentify/identify/

---

*[sql/scan.go, line 440 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e5D4ORSU7CGoQVOcj:-K8e5D4ORSU7CGoQVOck:-120635647) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/scan.go#L440)):*
the code that runs when `ok` is `true` is much smaller. can you rearrange this so it reads:
```
if ok {
...
continue
}
....
```

---

*[sql/scan.go, line 445 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e4tkMN7hpJFpq_RVr:-K8e4tkMN7hpJFpq_RVs:-1642877948) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/scan.go#L445)):*
you're mixing `b` and `v` throughout

---

*[sql/scan.go, line 450 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e4feGvlsujv_fwnEW:-K8e4feGvlsujv_fwnEX:-227492286) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/scan.go#L450)):*
nit: s/effectively //

---

*[sql/scan.go, line 460 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e4yq8nqfeqN3_1TIK:-K8e4yq8nqfeqN3_1TIL:791810835) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/scan.go#L460)):*
slightly confusing is the term "results" since you're really talking about the index's natural ordering.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Added some tests - I verified these don't pass without the rest of the change.

---

Review status: 6 of 12 files reviewed at latest revision, 13 unresolved discussions.

---

*[sql/distinct.go, line 37 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e-K310S1LQK9HGr_X:-K8kpx_ri1lnGMBgDua5:1255945559) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/distinct.go#L37)):*
orderingInfo cannot be nil (it's a struct)

orderingInfo should be able to represent an "empty" ordering without any special casing.

I agree the method name is weird though, I am changing to isEmpty

---

*[sql/distinct.go, line 40 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e2tND0x6FGbuyrACA:-K8koR1FGjPu0nzK4Go4:-881666568) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/distinct.go#L40)):*
The comment explained why it's ok to not do anything if the if condition is false. I rewrote the condition, hopefully it's clearer now.

---

*[sql/ordering.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8PqgnYlrZ1nMJCWH35:-K8lM7sJdhUHbdCW43lc:731255883) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L48)):*
I am ok with `exactMatchCols`, some code around this already uses the idea of "exact" (e.g. exactPrefix). Other names could be `singleValueCols` or `singleMatchCols`, let me know if any of those are better.

---

*[sql/ordering.go, line 71 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e-rLl2757t-8EQrol:-K8kqzrkHEs9HxmSw1OO:1356357415) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L71)):*
It seems like it doesn't belong in either columnOrdering or orderingInfo.. Where would you put it?

---

*[sql/ordering.go, line 79 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e0PwrJ-VxENXOOaOM:-K8krUqDbg86kwYyRD_i:1213567932) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L79)):*
I expanded the comment, hopefully it's clearer now.

---

*[sql/ordering_test.go, line 61 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8Psfl_zO_uutlKG-79:-K8kxCdw7mnSmh2AJqeL:-490100129) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering_test.go#L61)):*
Nice!

---

*[sql/values.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8Pu5IJLFELC77jElJN:-K8kpf3w-1DwwbYcH4u5:-1325955151) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/values.go#L89)):*
orderingInfo is a struct..

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-:-K8lLk9EGkarn61NqUsM:1868061649)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/ordering.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8PqgnYlrZ1nMJCWH35:-K8p62bEt_1NJEGJxjtd:1917943174) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L48)):*
I guess I prefer `singleMatchCols`, but I don't feel strongly.

---

*[sql/ordering.go, line 71 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e-rLl2757t-8EQrol:-K8p698YKzTdYNxwwpwD:1677485096) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L71)):*
I'd put it on orderingInfo, but I don't feel strongly.

---

*[sql/ordering.go, line 79 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8e0PwrJ-VxENXOOaOM:-K8p6EjvUUiQX5_rGUeZ:-1169850790) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L79)):*
The comment was clear before too, but I found it easier to reason about a comparison between two equivalent fields than the current code. I guess peter wasn't tripped up by this, so it's not a big deal.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/ordering.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8PqgnYlrZ1nMJCWH35:-K8tXSVWmLcMGmnQfBcf:-78341103) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L48)):*
Looking at this again, both `exactMatch` and `singleMatch` are a bit unsatisfying. Maybe something like `preSortedCols`. It's not necessary to change this again in this PR. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

petermattis commented:
LGTM

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-:-K8tYi4MQbSJ39idUrq5:676834233)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[sql/ordering.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8PqgnYlrZ1nMJCWH35:-K8tbrFEpwakYRaZOcDO:2058503817) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L48)):*
Hm. it's not that they are pre-sorted, or we wouldn't be able to match them to a desired ordering in whatever interleaving/order. The key is that there is a single value for that column in whatever set of results we are dealing with.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/ordering.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8PqgnYlrZ1nMJCWH35:-K8teGNq5rb00yl3OfGV:156095085) ([raw file](https://github.com/cockroachdb/cockroach/blob/1987ee23327941ea4f3ffd02f23fe6da1acfac82/sql/ordering.go#L48)):*
Yeah, they aren't actually sorted, but sorting isn't necessary. Let's leave this as-is for now. If we have a flash of inspiration we can revisit.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Reviewed 2 of 11 files at r1, 3 of 6 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[sql/ordering.go, line 26 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uAFGlUVj6x5lDEfI0:-K8uAFGmhMZQqSQl9lx-:-1836325537) ([raw file](https://github.com/cockroachdb/cockroach/blob/427ff38c43123b3d9f058afdd5ba68170bb70f7d/sql/ordering.go#L26)):*
how about using `encoding.Direction`?

---

*[sql/ordering.go, line 49 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uHXfgAziCnfmMJJ0K:-K8uHXfhZZLtxJbtE_l2:1278922144) ([raw file](https://github.com/cockroachdb/cockroach/blob/427ff38c43123b3d9f058afdd5ba68170bb70f7d/sql/ordering.go#L49)):*
maybe emphasize again that cols in `exactMatchCols` are *not* also part of this structure.

---

*[sql/scan.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uDjPKxZyOQDDVrjR4:-K8uDjPLaF59niKjXLpd:1755008024) ([raw file](https://github.com/cockroachdb/cockroach/blob/427ff38c43123b3d9f058afdd5ba68170bb70f7d/sql/scan.go#L410)):*
why both -1 and `ok`?

---

*[sql/scan.go, line 460 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uJbFpcVVnibg36b_Z:-K8uJbFqXZSZ8Hdnhq2C:-1975719591) ([raw file](https://github.com/cockroachdb/cockroach/blob/427ff38c43123b3d9f058afdd5ba68170bb70f7d/sql/scan.go#L460)):*
I'm not sure I follow this. What does a column being part of the output have to do with anything ordering-related? In your example below, what difference would it have made if `k` was also rendered?

---

*[sql/values.go, line 89 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uD90WDNRmNk1ETheW:-K8uD90WDNRmNk1ETheX:1936865102) ([raw file](https://github.com/cockroachdb/cockroach/blob/427ff38c43123b3d9f058afdd5ba68170bb70f7d/sql/values.go#L89)):*
when do we need to sort the rows in a `VALUES`? Would it make sense to see if they're already sorted (or pre-sort them) according to a primary key and return an `orderingInfo`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/ordering.go, line 26 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uAFGlUVj6x5lDEfI0:-K8uNGht3bXEejn4ftQs:-762256816) ([raw file](https://github.com/cockroachdb/cockroach/blob/427ff38c43123b3d9f058afdd5ba68170bb70f7d/sql/ordering.go#L26)):*
Sure, I will repaint the bike shed and see how it looks :)

---

*[sql/scan.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uDjPKxZyOQDDVrjR4:-K8uKz8fm6lruV0Yt2Qy:-1141744505) ([raw file](https://github.com/cockroachdb/cockroach/blob/427ff38c43123b3d9f058afdd5ba68170bb70f7d/sql/scan.go#L410)):*
I prefer using "ok" rather than special values.. but figured it would make sense to return a "bad" value in that case.

---

*[sql/scan.go, line 460 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uJbFpcVVnibg36b_Z:-K8uL7_Yn8s0pQVpIAaB:-58557148) ([raw file](https://github.com/cockroachdb/cockroach/blob/427ff38c43123b3d9f058afdd5ba68170bb70f7d/sql/scan.go#L460)):*
The difference would be that we would be able to expose the information that the results are ordered first by `k` then by `v`. Whatever layer is processing the results can decide if that ordering is useful or not. But since we are not exposing `k` we can't say anything. Ordered by `SOMETHING YOU CAN'T SEE` then by `v` is not useful.

---

*[sql/values.go, line 89 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uD90WDNRmNk1ETheW:-K8uMYrDx2t0i5FvwCHo:-904628511) ([raw file](https://github.com/cockroachdb/cockroach/blob/427ff38c43123b3d9f058afdd5ba68170bb70f7d/sql/values.go#L89)):*
This function is needed to implement planNode.  Detecting already sorted stuff might be useful in the future, e.g. for SELECT FROM (VALUES (..)) ..

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/scan.go, line 410 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uDjPKxZyOQDDVrjR4:-K8uOG9T0TZUC-0iWo1Q:1441332845) ([raw file](https://github.com/cockroachdb/cockroach/blob/427ff38c43123b3d9f058afdd5ba68170bb70f7d/sql/scan.go#L410)):*
Agree with using `ok`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Updated with Andrei's suggestions

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-:-K8uXcpQsLNd3G2oJCQ5:2109422250)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 8 of 8 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[sql/ordering.go, line 21 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8ugmrDCe9lEfuGvREn:-K8ugmrEoRMymxPS6MSy:366176335) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d034820ed861e2c1bc852f578cc323d5734d168/sql/ordering.go#L21)):*
nit: newline between these groups

---

*[sql/sort.go, line 113 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uh2yQsbkuPp0wHcYh:-K8uh2yQsbkuPp0wHcYi:1616124827) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d034820ed861e2c1bc852f578cc323d5734d168/sql/sort.go#L113)):*
what's all this about? why not use `o.Direction` here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[sql/sort.go, line 113 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uh2yQsbkuPp0wHcYh:-K8ujfwUZvenH3_IB4oS:-433073216) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d034820ed861e2c1bc852f578cc323d5734d168/sql/sort.go#L113)):*
Annoyingly o.Direction is parser.Direction which is different than encoding.Direction. I wish there was a common definition

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

tamird commented:
LGTM please squash before merging

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/sort.go, line 113 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8uh2yQsbkuPp0wHcYh:-K8ukvXB97nqHzbqmS-J:732186000) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d034820ed861e2c1bc852f578cc323d5734d168/sql/sort.go#L113)):*
Be the change! fine for now, though.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-:-K8ukzEtlwgF5vn2F3e3:-140316762)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/scan.go, line 434 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8v0ifjx9uczfDevqhY:-K8v0ifjx9uczfDevqhZ:1737705783) ([raw file](https://github.com/cockroachdb/cockroach/blob/00763a1c3ecb3bfd71b4eb2b369fe8882102b8d3/sql/scan.go#L434)):*
as we were discussing, I found it unclear what this function does. I think partly because the old comment had more hints than what's left of it (and what's left makes more harm than good).
How about making it take an index instead of the list of columns (I might have been the one that switched away from that? :S) and saying
```
computeOrdering returns the ordering of the columns produced by this scanNode assuming that index i would be scanned.
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-:-K8v5V4tdn-vjLD9HTts:676834233)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[sql/scan.go, line 434 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-K8v0ifjx9uczfDevqhY:-K8vIy4LHpW9rx3AdZWq:-1538605661) ([raw file](https://github.com/cockroachdb/cockroach/blob/08e88a4500e1ac2ed231414d461e7432c30454c5/sql/scan.go#L434)):*
Good idea, thanks. I also made it a standalone function, it should be clearer now.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
superb

---

Review status: 11 of 13 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3918#-:-K8ycKRb_dqgDfQRdlpa:-1237451380)*
<!-- Sent from Reviewable.io -->
